src/Test.UnitTests.Sarif/Visitors/SortingVisitorTests.cs

```diff
@@ -0,0 +1,313 @@
+                if (artifactLoc != null && artifactLoc.Index != -1)
+                {
+                    artifactLoc.Uri.Should().Be(artifacts[artifactLoc.Index].Location.Uri);
+                }
+            }
+        }
+
+        [Fact]
+        public void SortingVisitor_NullEmptyListTests()
+        {
+            // arrange
+            // create a log with all results have same value.
+            SarifLog sarifLog = new SarifLog
+            {
+                Runs = new[]
+                {
+                    new Run
+                    {
+                        Results = new[]
+                        {
+                            new Result
+                            {
+                                RuleId = "TESTRULE001"
+                            },
+                            new Result
+                            {
+                                RuleId = "TESTRULE001"
+                            },
+                            new Result
+                            {
+                                RuleId = "TESTRULE001"
+                            },
+                        },
+                    },
+                },
+            };
+
+            // setup results
+            IList<Result> results = sarifLog.Runs[0].Results;
+            results[0].Locations = new[] { new Location { Message = new Message { Text = "test location" } } };
+            results[1].Locations = null;
+            results[2].Locations = new List<Location>();
+
+            // act
+            SarifLog sortedLog = new SortingVisitor().VisitSarifLog(sarifLog);
+
+            // assert
+            // if collection elements all values are same expect a child list
+            // the order should depend on list values.
+            // expected order: null < empty < collection has element
+            results = sortedLog.Runs[0].Results;
+            results[0].Locations.Should().BeNull();
+            results[1].Locations.Should().BeEmpty();
+            results[2].Locations.Should().NotBeEmpty();
+        }
+
+        private static SarifLog CreateTestSarifLog(Random random)
+        {
+            SarifLog sarifLog = RandomSarifLogGenerator.GenerateSarifLogWithRuns(random, runCount: 1, resultCount: random.Next(100));
+
+            CreateCodeFlows(sarifLog.Runs?.First()?.Results, sarifLog.Runs?.First()?.Artifacts, random);
+
+            return sarifLog;
+        }
+
+        private static void CreateCodeFlows(IList<Result> results, IList<Artifact> artifacts, Random random)
+        {
+            if (results?.Any() != true || artifacts?.Any() != true)
+            {
+                return;
+            }
+
+            foreach (Result result in results)
+            {
+                result.CodeFlows = new[]
+                {
+                    new CodeFlow
+                    {
+                        ThreadFlows = new []
```

Comment:
>new [] [](http://example.com/codeflow?start=38&length=6)

should do a auto format code for all the files to keep it consistent, I tried no need space is the recommended one.

Reply:
which tool do you use? I ran dotnet format before commit, it didn't catch all of these additional spaces.

Reply:
in VS right click file -> code clean up with the default profile.